### PR TITLE
batches: SSBC animated library pane collapse

### PR DESCRIPTION
**Pull request by courier-new**
Upgrading reactstrap to a version that supports the vertical `Collapse` property is blocked on upgrading Bootstrap (see https://github.com/sourcegraph/sourcegraph/pull/29400) and will likely not be happening in the near future. Coupled with the fact that we are interested in introducing animation in a number of places in the new SSBC flow, including in the `WorkspacesPreview` and in page transitions, this felt like an appropriate place to introduce a custom animation instead as an example for future work.

We already had `react-spring` as a dev dependency for some reason, and react-spring appears to integrate fairly well with react-router for navigation transitions, so this felt like an apt choice.

The library pane adjacent to the editor now has a smooth transition:

https://user-images.githubusercontent.com/8942601/150623310-85a59ba4-b182-4f37-a038-f79c7f07d540.mov

The difficulty with animating things in/out of visibility is, of course, animating the `display` property, since we do want the hidden library pane to be removed from the DOM flow and made completely inaccessible while it is closed. Though pretty clunky, I was able to find a way to do this with react-spring's API *both* on-demand and sequentially with the other animations. There might be a better way to do this, but it was not clear from their docs or examples. 🤷‍♀️

### CODE REVIEW

File: client/wildcard/src/hooks/useMeasure.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    left: 0,
+    bottom: 0,
+    right: 0,
+}
+
+/**
+ * Custom hook for observing the size and position of an element using the Resize Observer
+ * API. Based on https://github.com/streamich/react-use/blob/master/src/useMeasure.ts
+ */
+export const useMeasure = <E extends Element = Element>(): [
```
Comment (courier-new) [inline client/wildcard/src/hooks/useMeasure.ts:20]: While I didn't end up needing this for this particular animation, I am going to need it for the others, so I left it in for now. 🙂

File: client/wildcard/src/hooks/index.ts [hunk 2]
```diff
@@ -1,16 +1,17 @@
...
 export { useLocalStorage } from './useLocalStorage'
 export { useMatchMedia } from './useMatchMedia'
+export { useMeasure } from './useMeasure'
 export { useEventObservable, useObservable } from './useObservable'
+export { useOffsetPagination } from './useOffsetPagination'
 export { useOnClickOutside } from './useOnClickOutside'
+export { useSearchParameters } from './useSearchParameters'
 export { useTimeoutManager } from './useTimeoutManager'
-export { useKeyboard } from './useKeyboard'
-export { useDeepMemo } from './useDeepMemo'
```
Comment (courier-new) [inline client/wildcard/src/hooks/index.ts:13]: This diff is just because I sorted these alphabetically.

**Comment by LawnGnome** [comment]
> Though pretty clunky, I was able to find a way to do this with react-spring's API _both_ on-demand and sequentially with the other animations. There might be a better way to do this, but it was not clear from their docs or examples.

I actually thought that was pretty clean, reading the documentation around the [API object return](https://react-spring.io/hooks/use-spring#or-pass-a-function-that-returns-values-and-update-using-the-api) and [async chains](https://react-spring.io/hooks/use-spring#async-chainsscripts).

Good job!

**Comment by courier-new** [comment]
Oh good, I'm glad it doesn't seem too clunky to you. 🙂 This is great validation that it's readable, too!

**Pull request merged**